File: Sources/LTXTextLayout.m
//
//  Created by Cyandev on 2022/5/8.
//  Copyright (c) 2024 ktiays. All rights reserved.
//

#import <CoreText/CoreText.h>

#import "LTXAttachment.h"
#import "LTXHighlightRegion+Private.h"
#import "LTXLineDrawingAction.h"
#import "LTXTextLayout.h"

static const CFRange kZeroRange = {0, 0};
static NSString *const kTruncationToken = @"\u2026";

static BOOL _hasHighlightAttributes(NSDictionary *attributes) {
    if ([attributes objectForKey:NSLinkAttributeName]) {
        return YES;
    }
    if ([attributes objectForKey:LTXAttachmentAttributeName]) {
        return YES;
    }
    return NO;
}

static _Nullable CTLineRef _createTruncatedLine(CTLineRef lastLine, NSAttributedString *attrString, CGFloat width) {
    // Determine the truncation token attributes. We use the last run's
    // attributes (if any) and strip some unsupported attribute keys.
    NSMutableDictionary *truncationTokenAttributes = [NSMutableDictionary dictionary];
    NSArray *lastLineGlyphRuns = (NSArray *) CTLineGetGlyphRuns(lastLine);
    CTRunRef lastGlyphRun = (__bridge CTRunRef) lastLineGlyphRuns.lastObject;
    if (lastGlyphRun) {
        NSDictionary *lastRunAttributes = (NSDictionary *) CTRunGetAttributes(lastGlyphRun);
#define _COPY_ATTR(key) \
    do { \
        id attr = lastRunAttributes[(key)]; \
        if (attr) { \
            truncationTokenAttributes[(key)] = attr; \
        } \
    } while (0)

        _COPY_ATTR(NSFontAttributeName);
        _COPY_ATTR(NSForegroundColorAttributeName);
        _COPY_ATTR(NSParagraphStyleAttributeName);

#undef _COPY_ATTR
    }

    NSAttributedString *truncationTokenString = [[NSAttributedString alloc] initWithString:kTruncationToken attributes:truncationTokenAttributes];
    CTLineRef truncationLine = CTLineCreateWithAttributedString((CFAttributedStringRef) truncationTokenString);

    CFRange lastLineStringCFRange = CTLineGetStringRange(lastLine);
    NSRange lastLineStringRange = {lastLineStringCFRange.location, lastLineStringCFRange.length};
    NSMutableAttributedString *lastLineString = [[attrString attributedSubstringFromRange:lastLineStringRange] mutableCopy];
    [lastLineString appendAttributedString:truncationTokenString];
    // Note: a new object allocated.
    lastLine = CTLineCreateWithAttributedString((CFAttributedStringRef) lastLineString);

    CTLineRef truncatedLine = CTLineCreateTruncatedLine(lastLine, width, kCTLineTruncationEnd, truncationLine);

    CFRelease(truncationLine);
    CFRelease(lastLine);

    return truncatedLine;
}

@interface LTXTextLayout ()

@property (nonatomic, readwrite, copy) NSAttributedString *attributedString;

@end

@implementation LTXTextLayout {
    CTFramesetterRef _framesetter;
    CTFrameRef _ctFrame;
    NSArray *_lines;
    NSMutableDictionary<NSNumber *, LTXHighlightRegion *> *_highlightRegions;
    NSMutableSet<LTXLineDrawingAction *> *_lineDrawingActions;
}

+ (instancetype)textLayoutWithAttributedString:(NSAttributedString *)attributedString {
    return [[self alloc] initWithAttributedString:attributedString];
}

- (instancetype)initWithAttributedString:(NSAttributedString *)attributedString {
    self = [super init];
    if (self) {
        _attributedString = attributedString;
        _framesetter = CTFramesetterCreateWithAttributedString((CFAttributedStringRef) _attributedString);
        _highlightRegions = [NSMutableDictionary dictionary];
    }
    return self;
}

- (void)dealloc {
    if (_framesetter) {
        CFRelease(_framesetter);
        _framesetter = NULL;
    }
    
    if (_ctFrame) {
        CFRelease(_ctFrame);
        _ctFrame = NULL;
    }
    
    [_highlightRegions removeAllObjects];
    _highlightRegions = nil;
    
    [_lineDrawingActions removeAllObjects];
    _lineDrawingActions = nil;
    
    _lines = nil;
    _attributedString = nil;
}

- (NSArray<LTXHighlightRegion *> *)highlightRegions {
    return [_highlightRegions allValues];
}

- (void)setContainerSize:(CGSize)containerSize {
    _containerSize = containerSize;
    [self _generateLayout];
}

- (void)invalidateLayout {
    [self _generateLayout];
}

- (CGSize)suggestContainerSizeWithSize:(CGSize)size {
    return CTFramesetterSuggestFrameSizeWithConstraints(_framesetter, kZeroRange, NULL, size, NULL);
}

- (void)drawInContext:(CGContextRef)context {
    if (!_lineDrawingActions) {
        _lineDrawingActions = [NSMutableSet set];
    }
    [_lineDrawingActions removeAllObjects];

    CGContextSaveGState(context);

    // Flip the rendering coordinate.
    CGSize containerSize = _containerSize;
    CGContextTranslateCTM(context, 0, containerSize.height);
    CGContextScaleCTM(context, 1, -1);

    CTFrameDraw(_ctFrame, context);

    [self _enumerateLinesUsingBlock:^(CTLineRef line, NSUInteger idx, CGPoint lineOrigin) {
        NSArray *glyphRuns = (NSArray *) CTLineGetGlyphRuns(line);

        [glyphRuns enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
            CTRunRef glyphRun = (__bridge CTRunRef) obj;

            NSDictionary *attributes = (NSDictionary *) CTRunGetAttributes(glyphRun);
            LTXLineDrawingAction *action;
            if ((action = [attributes objectForKey:LTXLineDrawingCallbackName])) {
                if ([_lineDrawingActions containsObject:action]) {
                    return;
                }
                CGContextSaveGState(context);
                action.action(context, line, lineOrigin);
                CGContextRestoreGState(context);
                if (action.performOncePerAttribute) {
                    [_lineDrawingActions addObject:action];
                }
            }
        }];
    }];

    CGContextRestoreGState(context);
}

- (void)updateHighlightRegionsWithContext:(CGContextRef)context {
    [_highlightRegions removeAllObjects];
    [self _extractHighlightRegionsWithContext:context];
}

- (void)_generateLayout {
    if (_ctFrame) {
        CFRelease(_ctFrame);
        _ctFrame = nil;
        _lines = nil;
    }

    CGRect containerBounds = {.origin = CGPointZero, .size = _containerSize};
    CGPathRef containerPath = CGPathCreateWithRect(containerBounds, NULL);
    _ctFrame = CTFramesetterCreateFrame(_framesetter, kZeroRange, containerPath, NULL);
    _lines = (NSArray *) CTFrameGetLines(_ctFrame);

    // Handle line truncation.
    CFRange visibleRange = CTFrameGetVisibleStringRange(_ctFrame);
    if (visibleRange.length == _attributedString.length || _lines.count == 0) {
        CGPathRelease(containerPath);
        return;
    }

    CTLineRef lastLine = (__bridge CTLineRef) _lines.lastObject;
    CTLineRef truncatedLine = _createTruncatedLine(lastLine, _attributedString, CGRectGetWidth(containerBounds));

    if (truncatedLine) {
        NSMutableArray *lines = [_lines mutableCopy];
        lines[lines.count - 1] = (__bridge_transfer id) truncatedLine;
        _lines = [lines copy];
    }
    CGPathRelease(containerPath);
}

- (void)_extractHighlightRegionsWithContext:(CGContextRef)context {
    __auto_type attributedString = _attributedString;
    __auto_type highlightRegions = _highlightRegions;

    [self _enumerateLinesUsingBlock:^(CTLineRef line, NSUInteger idx, CGPoint lineOrigin) {
        NSArray *glyphRuns = (NSArray *) CTLineGetGlyphRuns(line);

        [glyphRuns enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
            CTRunRef glyphRun = (__bridge CTRunRef) obj;
            NSDictionary *attributes = (NSDictionary *) CTRunGetAttributes(glyphRun);
            if (!_hasHighlightAttributes(attributes)) {
                return;
            }

            NSRange effectiveRange;
            CFRange cfStringRange = CTRunGetStringRange(glyphRun);
            NSRange stringRange = {cfStringRange.location, cfStringRange.length};
            [attributedString attributesAtIndex:stringRange.location effectiveRange:&effectiveRange];

            // Merge all highlight regions by attribute groups (with the same effective range).
            LTXHighlightRegion *highlightRegion = highlightRegions[@(effectiveRange.location)];
            if (!highlightRegion) {
                highlightRegion = [[LTXHighlightRegion alloc] initWithAttributes:attributes stringRange:stringRange];
                highlightRegions[@(effectiveRange.location)] = highlightRegion;
            }

            CGRect runBounds = CTRunGetImageBounds(glyphRun, context, kZeroRange);
            LTXAttachment *attachment;
            if ((attachment = [attributes objectForKey:LTXAttachmentAttributeName])) {
                runBounds.size = attachment.size;
                runBounds.origin.y -= attachment.size.height * 0.1;
            }
            runBounds.origin.x += lineOrigin.x;
            runBounds.origin.y += lineOrigin.y;
            [highlightRegion addRect:runBounds];
        }];
    }];
}

- (void)_enumerateLinesUsingBlock:(void (^)(CTLineRef line, NSUInteger idx, CGPoint origin))block {
    NSArray *lines = _lines;
    CGPoint *lineOrigins = malloc(sizeof(CGPoint) * lines.count);
    CTFrameGetLineOrigins(_ctFrame, kZeroRange, lineOrigins);

    [lines enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
        CTLineRef line = (__bridge CTLineRef) obj;
        CGPoint lineOrigin = lineOrigins[idx];
        block(line, idx, lineOrigin);
    }];

    free(lineOrigins);
}

@end

File: Sources/Litext/LTXTextLayout.swift
//
//  Created by Lakr233 & Helixform on 2025/2/18.
//  Copyright (c) 2025 Litext Team. All rights reserved.
//

import CoreGraphics
import CoreText
import Foundation
import QuartzCore

private let kTruncationToken = "\u{2026}"

private func _hasHighlightAttributes(_ attributes: [NSAttributedString.Key: Any]) -> Bool {
    if attributes[.link] != nil {
        return true
    }
    if attributes[LTXAttachmentAttributeName] != nil {
        return true
    }
    return false
}

private func _createTruncatedLine(lastLine: CTLine, attrString: NSAttributedString, width: CGFloat) -> CTLine? {
    var truncationTokenAttributes: [NSAttributedString.Key: Any] = [:]
    let lastLineGlyphRuns = CTLineGetGlyphRuns(lastLine) as NSArray
    if let lastGlyphRun = lastLineGlyphRuns.lastObject as! CTRun? {
        let lastRunAttributes = CTRunGetAttributes(lastGlyphRun) as! [NSAttributedString.Key: Any]

        if let font = lastRunAttributes[.font] {
            truncationTokenAttributes[.font] = font
        }
        if let foregroundColor = lastRunAttributes[.foregroundColor] {
            truncationTokenAttributes[.foregroundColor] = foregroundColor
        }
        if let paragraphStyle = lastRunAttributes[.paragraphStyle] {
            truncationTokenAttributes[.paragraphStyle] = paragraphStyle
        }
    }

    let truncationTokenString = NSAttributedString(string: kTruncationToken, attributes: truncationTokenAttributes)
    let truncationLine = CTLineCreateWithAttributedString(truncationTokenString)

    let lastLineStringRange = CTLineGetStringRange(lastLine)
    let nsRange = NSRange(location: lastLineStringRange.location, length: lastLineStringRange.length)
    let lastLineString = NSMutableAttributedString(attributedString: attrString.attributedSubstring(from: nsRange))
    lastLineString.append(truncationTokenString)
    let newLastLine = CTLineCreateWithAttributedString(lastLineString)

    let truncatedLine = CTLineCreateTruncatedLine(newLastLine, width, .end, truncationLine)

    return truncatedLine
}

public class LTXTextLayout: NSObject {
    public private(set) var attributedString: NSAttributedString
    public var highlightRegions: [LTXHighlightRegion] {
        Array(_highlightRegions.values)
    }

    public var containerSize: CGSize {
        didSet {
            generateLayout()
        }
    }

    private var framesetter: CTFramesetter
    private var ctFrame: CTFrame?
    private var lines: [CTLine]?
    private var highlightRegionsByLocation: [Int: LTXHighlightRegion] = [:]
    private var lineDrawingActions: Set<LTXLineDrawingAction> = []

    public class func textLayout(withAttributedString attributedString: NSAttributedString) -> LTXTextLayout {
        LTXTextLayout(attributedString: attributedString)
    }

    public init(attributedString: NSAttributedString) {
        self.attributedString = attributedString
        containerSize = .zero
        framesetter = CTFramesetterCreateWithAttributedString(attributedString)
        _highlightRegions = [:]
        super.init()
    }

    deinit {}

    private var _highlightRegions: [Int: LTXHighlightRegion]

    public func invalidateLayout() {
        generateLayout()
    }

    public func suggestContainerSize(withSize size: CGSize) -> CGSize {
        CTFramesetterSuggestFrameSizeWithConstraints(
            framesetter,
            CFRange(location: 0, length: 0),
            nil,
            size,
            nil
        )
    }

    public func draw(in context: CGContext) {
        lineDrawingActions.removeAll()

        context.saveGState()

        context.setAllowsAntialiasing(true)
        context.setShouldSmoothFonts(true)

        context.translateBy(x: 0, y: containerSize.height)
        context.scaleBy(x: 1, y: -1)

        if let ctFrame { CTFrameDraw(ctFrame, context) }

        enumerateLines { line, _, lineOrigin in
            let glyphRuns = CTLineGetGlyphRuns(line) as NSArray

            for i in 0 ..< glyphRuns.count {
                guard let glyphRun = glyphRuns[i] as! CTRun? else { continue }

                let attributes = CTRunGetAttributes(glyphRun) as! [NSAttributedString.Key: Any]
                if let action = attributes[LTXLineDrawingCallbackName] as? LTXLineDrawingAction {
                    if self.lineDrawingActions.contains(action) {
                        continue
                    }
                    context.saveGState()
                    action.action(context, line, lineOrigin)
                    context.restoreGState()
                    if action.performOncePerAttribute {
                        self.lineDrawingActions.insert(action)
                    }
                }
            }
        }

        context.restoreGState()
    }

    public func updateHighlightRegions(with context: CGContext) {
        _highlightRegions.removeAll()
        extractHighlightRegions(with: context)
    }

    // MARK: - Private Methods

    private func generateLayout() {
        lines = nil

        let containerBounds = CGRect(origin: .zero, size: containerSize)
        let containerPath = CGPath(rect: containerBounds, transform: nil)
        ctFrame = CTFramesetterCreateFrame(framesetter, CFRange(location: 0, length: 0), containerPath, nil)

        if let ctFrame {
            lines = CTFrameGetLines(ctFrame) as? [CTLine]
        }

        if let lines, let ctFrame {
            let visibleRange = CTFrameGetVisibleStringRange(ctFrame)
            if visibleRange.length == attributedString.length || lines.isEmpty {
                return
            }

            if let lastLine = lines.last, let truncatedLine = _createTruncatedLine(lastLine: lastLine, attrString: attributedString, width: containerBounds.width) {
                var newLines = lines
                newLines[newLines.count - 1] = truncatedLine
                self.lines = newLines
            }
        }
    }

    private func extractHighlightRegions(with context: CGContext) {
        enumerateLines { line, _, lineOrigin in
            let glyphRuns = CTLineGetGlyphRuns(line) as NSArray

            for i in 0 ..< glyphRuns.count {
                guard let glyphRun = glyphRuns[i] as! CTRun? else { continue }

                let attributes = CTRunGetAttributes(glyphRun) as! [NSAttributedString.Key: Any]
                if !_hasHighlightAttributes(attributes) {
                    continue
                }

                let cfStringRange = CTRunGetStringRange(glyphRun)
                let stringRange = NSRange(location: cfStringRange.location, length: cfStringRange.length)

                var effectiveRange = NSRange()
                _ = self.attributedString.attributes(at: stringRange.location, effectiveRange: &effectiveRange)

                let highlightRegion: LTXHighlightRegion
                if let existingRegion = self._highlightRegions[effectiveRange.location] {
                    highlightRegion = existingRegion
                } else {
                    highlightRegion = LTXHighlightRegion(attributes: attributes, stringRange: stringRange)
                    self._highlightRegions[effectiveRange.location] = highlightRegion
                }

                var runBounds = CTRunGetImageBounds(glyphRun, context, CFRange(location: 0, length: 0))

                if let attachment = attributes[LTXAttachmentAttributeName] as? LTXAttachment {
                    runBounds.size = attachment.size
                    runBounds.origin.y -= attachment.size.height * 0.1
                }

                runBounds.origin.x += lineOrigin.x
                runBounds.origin.y += lineOrigin.y
                highlightRegion.addRect(runBounds)
            }
        }
    }

    private func enumerateLines(using block: (CTLine, Int, CGPoint) -> Void) {
        guard let lines, let ctFrame else { return }

        let lineCount = lines.count
        var lineOrigins = [CGPoint](repeating: .zero, count: lineCount)
        CTFrameGetLineOrigins(ctFrame, CFRange(location: 0, length: 0), &lineOrigins)

        for i in 0 ..< lineCount {
            let line = lines[i]
            let origin = lineOrigins[i]
            block(line, i, origin)
        }
    }
}
